Updated Comments likes UI changes to be instant

ronaldlangeveld opened the pull request:
ref PLG-288

- Implemented instant UI updates for comment likes/unlikes using optimistic rendering.
- Enhanced error handling reverts state on API failure, ensuring consistency. 
- Added new test helper to mock failed API requests, needed to test revert state handling.

ronaldlangeveld committed:
Revert "Removed blocking tests (#21747)"

This reverts commit 72125f352b9603fd39ce9c709350d4abada951ef.

ronaldlangeveld committed:
[wip] Make comments like instant

ref PLG-288

- spike implementation of handling UI state changes while the API
  request runs in the background

ronaldlangeveld committed:
Revert "Revert "Removed blocking tests (#21747)""

This reverts commit 858f8c3f1640c1b435ae2119ffbc95ce44168fb6.

ronaldlangeveld committed:
Added test helper to set failures

ronaldlangeveld committed:
Added test

ronaldlangeveld committed:
Updated state name to be more readible

ronaldlangeveld committed:
Cleanup and added additional test

kevinansfield commented on apps/comments-ui/src/actions.ts:
```diff
@@ -186,9 +186,29 @@ async function showComment({state, api, data: comment}: {state: EditableAppConte
+            commentLikeLoadingId: null
+        };
+    } catch (err) {
+        // if error we revert the state
+        return {
+            comments: commentsState,
+            commentLikeLoadingId: null
+        };
```
This seems dangerous as other state changes could have happened if there was a long delay before the endpoint eventually errored. I would expect a rollback to only occur for the liked status of the comment under change, would that make things considerably more difficult?

kevinansfield commented on apps/comments-ui/src/actions.ts:
```diff
@@ -222,19 +242,33 @@ async function likeComment({state, api, data: comment}: {state: EditableAppConte
-    return {};
+    try {
+        await api.comments.unlike({comment: {id: comment.id}});
+        return {
+            commentUnlikeLoading: null
+        };
```
Do we need this? It feels clunky.

Can we keep `likeComment` as it was where it still awaits the API response but first dispatches an action that does an immediate state update? That way the component that dispatches `likeComment` can await as normal and manage the button disabling with a typical wrapper around an async action but the state still updates immediately in the background to show the like/unlike status

kevinansfield replied:
Pseudo-code:

```
async function likeComment(id) {
    dispatchAction('updateCommentLikeState', {id, liked: true});
    try {
        await api.likeComment(id);
    } catch () {
        dispatchAction('updateCommentLikeState', {id, liked: false});
    }
}
```

ronaldlangeveld committed:
Merge branch 'main' into make-comments-likes-instant

ronaldlangeveld committed:
Changed code based on feedback

ref https://github.com/TryGhost/Ghost/pull/21861#discussion_r1881701985

- simplified disabled logic to be within the component instead of
  globally.
- added state change in a single dispatch query that can handle like
  / unlike as well as fallback handling.
- updated test

ronaldlangeveld committed:
Cleanup

ronaldlangeveld committed:
Removed unused variables
